File: android/supercamera/src/main/java/com/dcamera/render/PreviewRenderBridge.kt
package com.dcamera.render

class PreviewRenderBridge(
    private val renderPipeline: RenderPipelinePort = NoOpRenderPipelinePort(),
) : PreviewBridgePort {
    private var surfaceAttached: Boolean = false
    private var frameCounter: Long = 0

    override fun attachSurface(surfaceHandle: Long): Boolean {
        if (surfaceHandle == 0L) return false
        surfaceAttached = true
        frameCounter = 0
        return true
    }

    override fun detachSurface() {
        surfaceAttached = false
    }

    fun submitFrame(frame: PreviewFrame): Boolean {
        if (!surfaceAttached) return false
        if (frame.width <= 0 || frame.height <= 0) return false
        if (frame.timestampNs < 0) return false
        if (!renderPipeline.submitFrame(frame)) return false
        frameCounter++
        return true
    }

    fun frameCount(): Long = frameCounter
}

File: ios/SuperCameraKit/Sources/SuperCameraKit/PreviewRenderBridge.swift
import Foundation

public final class PreviewRenderBridge {
    private var surfaceAttached: Bool = false
    public private(set) var frameCount: Int = 0

    public init() {}

    @discardableResult
    public func attach(surfaceHandle: UInt64) -> Bool {
        guard surfaceHandle != 0 else { return false }
        surfaceAttached = true
        frameCount = 0
        return true
    }

    public func detach() {
        surfaceAttached = false
    }

    @discardableResult
    public func submitFrame(frameHandle: UInt64, timestampNs: Int64) -> Bool {
        guard surfaceAttached, frameHandle != 0, timestampNs >= 0 else { return false }
        frameCount += 1
        return true
    }
}
